File: xr18/lib/src/main/java/com/mixer/xr18/lib/domain/model/ChannelState.java
package com.mixer.xr18.lib.domain.model;

/**
 * Complete state for one mixer input channel (CH01-CH16).
 */
public class ChannelState {
    public static final float PAN_LEFT = 0f;
    public static final float PAN_CENTER = 0.5f;
    public static final float PAN_RIGHT = 1f;

    public int channelNumber;         // 1-16
    public float fader;               // 0.0–1.0 (linear)
    public float faderDb;             // dB equivalent
    public boolean muted;
    public float pan;                 // 0.0 (L) – 0.5 (C) – 1.0 (R)
    public float preampGain;          // raw float 0.0-1.0
    public float preampGainDb;         // gain in dB (raw * 72 - 12, range -12 to +60)
    public float meter;                 // 0.0–1.0 (linear, from /meters blob)
    public float meterDb;               // meter in dB
    public boolean eqEnabled;
    public EqBands eqBands;

    public ChannelState(int channelNumber) {
        this.channelNumber = channelNumber;
        this.fader = 0.707f;
        this.faderDb = 0f;
        this.muted = false;
        this.pan = 0.5f;
        this.preampGain = 0f;
        this.preampGainDb = -12f;
        this.meter = 0f;
        this.meterDb = -96f;
        this.eqEnabled = false;
        this.eqBands = new EqBands();
    }

    public static float faderToDb(float linear) {
        // Empirical formula from simultaneous data with X AIR Edit / Mixing Station
        // UI_dB = 58.1 * raw - 43.5 (linear fit for CH01=-26.4, CH03=0.0)
        if (linear <= 0f) return Float.NEGATIVE_INFINITY;
        return (float) (58.1f * linear - 43.5f);
    }

    public String faderPercent() {
        return String.format("%.1f%%", fader * 100f);
    }

    public static float gainToDb(float rawGain) {
        // XR18 headamp gain: raw 0.0-1.0 maps to -12 to +60 dB
        return rawGain * 72f - 12f;
    }

    /** Convert meter value (16-bit signed int, resolution 1/256 dB) to linear 0-1. */
    public static float meterValueToLinear(short meterValue) {
        // dB = meterValue / 256.0
        // linear = 10^(dB/20)
        // XR18 meter range: -96dB (meterValue=-24576) to +12dB (meterValue=+3072)
        // Clamp only below -96 dB to avoid displaying extremely weak signals as 0
        if (meterValue <= -24576) return 0f;  // below -96 dB → silent
        double db = meterValue / 256.0;
        return (float) Math.pow(10.0, db / 20.0);
    }

    /** Convert meter value (16-bit signed int) to dB. */
    public static float meterValueToDb(short meterValue) {
        // Clamp: XR18 meters valid range is -96dB to +12dB
        if (meterValue < -24576) meterValue = -24576;
        if (meterValue > 3072) meterValue = 3072;
        float db = meterValue / 256.0f;
        // NEVER display more than +100dB (hard clamp for UI safety)
        if (db > 100f) db = 100f;
        return db;
    }

    public String preampGainDbString() {
        if (preampGainDb <= -90f) return "–∞ dB";
        return String.format("%+.1f dB", preampGainDb);
    }

    public String meterDbString() {
        if (meterDb <= -96f) return "–∞ dB";
        return String.format("%+.1f dB", meterDb);
    }

    public String faderDbString() {
        if (fader <= 0.001f) return "–∞ dB";
        return String.format("%+.1f dB", faderDb);
    }

    public String panString() {
        return panToString(this.pan);
    }

    public static String panToString(float pan) {
        if (pan <= 0f) return "L100";
        if (pan < 0.25f) return "L" + (int) (100 * (1 - pan * 4));
        if (pan < 0.5f) return "L" + (int) (50 * (0.5f - pan) * 4);
        if (pan == 0.5f) return "C";
        if (pan < 0.75f) return "R" + (int) (50 * (pan - 0.5f) * 4);
        if (pan < 1f) return "R" + (int) (100 * (pan - 0.25f) * 4);
        return "R100";
    }

    public ChannelState withFader(float v) { this.fader = v; this.faderDb = faderToDb(v); return this; }
    public ChannelState withMuted(boolean v) { this.muted = v; return this; }
    public ChannelState withPan(float v) { this.pan = v; return this; }
    public ChannelState withPreampGain(float v) { this.preampGain = v; this.preampGainDb = gainToDb(v); return this; }
    public ChannelState withEqEnabled(boolean v) { this.eqEnabled = v; return this; }
    public ChannelState withEqBands(EqBands v) { this.eqBands = v; return this; }
}

File: xr18/lib/src/main/java/com/mixer/xr18/lib/domain/model/ChannelState.kt
package com.mixer.xr18.lib.domain.model

/**
 * EQ band gain values for a single channel.
 */
data class EqBands(
    val band1: Float = 0f,
    val band2: Float = 0f,
    val band3: Float = 0f,
    val band4: Float = 0f
)

/**
 * Complete state for one mixer input channel (CH01-CH16).
 */
data class ChannelState(
    val channelNumber: Int,          // 1-16
    val fader: Float = 0.707f,       // 0.0–1.0 (linear)
    val faderDb: Float = 0f,         // dB equivalent
    val muted: Boolean = false,
    val pan: Float = 0.5f,          // 0.0 (L) – 0.5 (C) – 1.0 (R)
    val preampGain: Float = 0f,     // dB
    val eqEnabled: Boolean = false,
    val eqBands: EqBands = EqBands()
) {
    companion object {
        const val PAN_LEFT = 0f
        const val PAN_CENTER = 0.5f
        const val PAN_RIGHT = 1f

        fun panToString(pan: Float): String = when {
            pan <= 0f          -> "L100"
            pan < 0.25f        -> "L$((100 * (1 - pan * 4)).toInt())"
            pan < 0.5f         -> "L$((50 * (0.5f - pan) * 4).toInt())"
            pan == 0.5f        -> "C"
            pan < 0.75f        -> "R$((50 * (pan - 0.5f) * 4).toInt())"
            pan < 1f           -> "R$((100 * (pan - 0.25f * 4)).toInt())"
            else               -> "R100"
        }

        /**
         * Convert linear fader (0–1) to dB.
         * 1.0 = +10 dB, 0.707 ≈ 0 dB, 0.0 = –∞ dB
         */
        fun faderToDb(linear: Float): Float {
            if (linear <= 0f) return Float.NEGATIVE_INFINITY
            return 20f * kotlin.math.log10(linear) + 10f
        }
    }

    fun faderPercent(): String = "%.1f%%".format(fader * 100)
    fun faderDbString(): String {
        val db = faderDb
        return if (db <= -90f) "–∞ dB" else "%+.1f dB".format(db)
    }
    fun panString(): String = panToString(pan)
}
